File: Two Pointers/2824. Count Pairs Whose Sum is Less than Target/2824. Count Pairs Whose Sum is Less than Target.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var countPairs = function (nums, target) {
  nums.sort((a, b) => a - b);
  let l = 0;
  let r = nums.length - 1;
  let pairs = 0;

  while (l < r) {
    if (nums[l] + nums[r] < target) {
      pairs += r - l;
      l++;
    } else {
      r--;
    }
  }

  return pairs;
};

// Time: O(n log n)
// Space: O(1)

File: Two Pointers/2824. Count Pairs Whose Sum is Less than Target/2824. Count Pairs Whose Sum is Less than Target.py
class Solution:
    def countPairs(self, nums: List[int], target: int) -> int:
        nums.sort()
        l = 0
        r = len(nums) - 1
        pairs = 0

        while l < r:
            if nums[l] + nums[r] < target:
                pairs += r - l
                l += 1
            else:
                r -= 1

        return pairs

# Time: O(n log n)
# Space: O(1)